overlord/ifacestate/ifacemgr.go

```diff
@@ -288,6 +289,96 @@ func (m *InterfaceManager) ConnectionStates() (connStateByRef map[string]Connect
 	return ConnectionStates(m.state)
 }
 
+// ResolveDisconnect resolves potentially missing plug or slot names and
+// returns a list of fully populated connection references that can be
+// disconnected.
+//
+// It can be used in two different ways:
+// 1: snap disconnect <snap>:<plug> <snap>:<slot>
+// 2: snap disconnect <snap>:<plug or slot>
+//
+// In the first case the referenced plug and slot must be connected.  In the
+// second case any matching connection are returned but it is not an error if
+// there are no connections.
+//
+// In both cases the snap name can be omitted to implicitly refer to the core
+// snap. If there's no core snap it is simply assumed to be called "core" to
+// provide consistent error messages.
+func (m *InterfaceManager) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slotName string) ([]*interfaces.ConnRef, error) {
+	connected := func(plugSn, plug, slotSn, slot string) (bool, error) {
+		_, err := m.repo.Connection(&interfaces.ConnRef{
+			PlugRef: interfaces.PlugRef{Snap: plugSn, Name: plug},
+			SlotRef: interfaces.SlotRef{Snap: slotSn, Name: slot},
+		})
+		if _, notConnected := err.(*interfaces.NotConnectedError); notConnected {
+			return false, nil
+		}
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+
+	connectedPlugOrSlot := func(snapName, plugOrSlotName string) ([]*interfaces.ConnRef, error) {
+		return m.repo.Connected(snapName, plugOrSlotName)
+	}
+
+	coreSnapName := SystemSnapName()
+	if coreSnapName == "" {
```

Comment:
it could happen with guess but can this happen with SystemSnapName ?

Reply:
I don't think so. Although this made me realize we may have another problem. Repository would dynamically adjust to pick the right system snap name as snaps came and went. The code in the interface manager makes a one-time decision on snapd startup. I'm not sure this is correct with the new responsibilities. In the past if was only used for core configuration to handle things like "system" alias. Here it has more responsibility.

Reply:
What happens in practice at the moment is that if we go from:

* no snapd to snapd
* or from core to snapd

there is a snapd restart involved,  so we do not meaningfully operate with the wrong value. This is true atm even for first boot installation of snapd.

Relevant code is:

https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1316
https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1351

Reply:
> it could happen with guess but can this happen with SystemSnapName ?

Will address in a followup.